Multi-file running changelog strategy

SevaZhukov opened the pull request:
Our current running changelog strategy is a reason for merge conflicts in every 1-3 PR. The new Multi-file running changelog strategy will solve this problem. How will it work you can read [here](https://github.com/mapbox/mapbox-navigation-android/blob/NAVSDK-646/changelog/README.md).

@VysotskiVadim already tried to run Multi-file running changelog strategy in #5354 but he was faced with the need for maintaining a lot of code on CI for assembling changes. Now we can delegate assembling to the release train app.

- [x] create README.md about the new strategy
- [x] move unreleased features to json files
- [x] refactor CI scripts with verification changelog
- [x] fix tests of these scripts
- [ ] refactor the train's changelog steps

dzinad commented on changelog/unreleased/6540.json:
```diff
@@ -0,0 +1,6 @@
+[
+  {
+    "type": "Bug fixes and improvements",
+    "title": "Fixed approaches list update in `RouteOptionsUpdater`(uses for reroute). It was putting to the origin approach corresponding approach from legacy approach list."
```
Will the PR links be added automatically?

SevaZhukov replied:
Yes

dzinad commented on changelog/unreleased/6540.json:
```diff
@@ -0,0 +1,6 @@
+[
+  {
+    "type": "Bug fixes and improvements",
```
I see how we can easily make a typo here and get a weird header. Can we make it fool-proof? For example:
We only have a finite number of headers. Features, Bug fixes and improvements, Known issues.
Can we create directories for them? And then what's left is place a change entry as a plain string.
So for this example it would be:
1) find directory "changelog/unreleased/Bug fixes and improvements"
2) Create a file "6540.json"
3) The file contents:
```
[
"Fixed approaches list update in `RouteOptionsUpdater`(uses for reroute). It was putting to the origin approach corresponding approach from legacy approach list."
]
```
If there are multiple entries, we just add a second array item.

What I want to do here is to not use any json keys. Because it creates room for errors as well as typing "Bug fixes and improvements" every time.
I think an array of string would be sufficient, WDYT?

VysotskiVadim replied:
If the only field that left in file is title, maybe we can replace `json` by `md`?

SevaZhukov replied:
I think it's a good point. In our case we have only 2 main parts: features and bugfixes, we can move type to path and make changelog file as md file

tomaszrybakiewicz commented:
Few things to keep in mind. Changelog entries may have:
- code snippets
  ```kotlin
  fun foo(bar: String) = "lorem ipsum dolor sit ammet..."
  ```
- foldable sections (`<details><summary></summary></details>`)
  <details><summary>Open for cake</summary>Cake is a lie!!!</details>
- **bold**, _italic_ and `inline code` text style
- emojis :warning:
- images

kmadsen commented on changelog/example.json:
```diff
@@ -0,0 +1,6 @@
+  {
+    "type": "Features",
+    "title": "Description of changes"
+  }
```
```suggestion
  },
  {
    "type": "Bug fixes and improvements",
    "title": "Description of changes"
  }
```

thinking it's good to have the example include `type` possibilities

kmadsen commented:
Spent some time thinking/discussing and am sharing that I'm in favor. Great idea, and I like how it is something that GitLab had an issue with and came to a similar conclusion. They liked the solution enough to write a blog post https://about.gitlab.com/blog/2018/07/03/solving-gitlabs-changelog-conflict-crisis/

I have some minor opinions on the details, but overall I'm in favor. Cleaner solution than what we're doing now, the issue will be fixed, and it can be expanded for more improvements 👍

kmadsen commented on changelog/README.md:
````diff
@@ -0,0 +1,59 @@
+- Mapbox Java `v6.8.0` ([release notes](https://github.com/mapbox/mapbox-java/releases/tag/v6.8.0))
+- Mapbox Android Core `v5.0.2` ([release notes](https://github.com/mapbox/mapbox-events-android/releases/tag/core-5.0.2))
+```
+* write the changelog to the `CHANGELOG.md` file
````
capturing @tomaszrybakiewicz comment.

Instead of generating CHANGELOG during release, do this on PR merge.

I'm also +1'ing that. It is nice to see a running changelog, instead of needing to dig through each new file created. 

Considering that may re-introduce the merge conflict issue, maybe it could be part of a script where we can say `sh scripts/show-unreleased-changelog.sh`?

kmadsen replied:
an idea is to have a file that represent the latest. every time a new pull request is merged the other one is deleted. 

```
pr 01 is merged
01-changelog.md

pr 02 is merged
01-changelog.md is deleted
02-changelog.md is created
```

not the best solution considering that the unreleased changelog link would have a short life https://github.com/mapbox/mapbox-navigation-android/blob/main/unreleased/01-changelog.md

VysotskiVadim replied:
Do we want to keep current not released version of running changelog publicly available? 
If not, maybe we can generate current changelog by request inside the release train app?

RingerJK replied:
>After merging a PR,

it means that changelog must be merged separately, in another commit, out of PR merge. It seems risky: make it in 1 step (aka transaction) safer, than drop to a few

RingerJK replied:
also, this way we will have 2x commits in the branch (1 is changed + 1 changelog)

SevaZhukov replied:
Do we really need unreleased changelog in one file before release?

tomaszrybakiewicz replied:
> it means that changelog must be merged separately, in another commit, out of PR merge. It seems risky: make it in 1 step (aka transaction) safer, than drop to a few

@RingerJK It's as risky as Release Train or CI release workflows making additional commits.

> also, this way we will have 2x commits in the branch (1 is changed + 1 changelog)

And why is that a problem?

SevaZhukov replied:
I think we can add one more job for the main branch to generate a running changelog file. This job will be executed after every merge

VysotskiVadim commented on changelog/README.md:
```diff
@@ -0,0 +1,59 @@
+[
+  {
+    "type": "Features",
+    "title": "Description of changes"
```
is title a string that contains markdown?

VysotskiVadim commented on changelog/README.md:
```diff
@@ -0,0 +1,59 @@
+
+To avoid merge conflicts in the CHANGELOG.md file we accepted the multi-file running changelog strategy.
+
+To follow this strategy you should create a `changelog/unreleased/${PR_NUMBER}.json` file for every PR like:
```
we usually add many lines during bump of NN and common dependencies. Does it mean that we will have to add as many files as many line we want to add during update of NN?

SevaZhukov replied:
yeah, if we move features/bugfixes type to the path we can add files with additional info like this in the unreleased directory and structure will be like:
```
changelog
--unreleased
----features
------feature_1.md
----bugfixes
------bugfix_1.md
----other_change_1.md
----other_change_2.md
```

VysotskiVadim commented:
Should we let to insert sections like [this](https://github.com/mapbox/mapbox-navigation-android/blob/main/CHANGELOG.md#known-issues-warning) or it should be done manually during release?

RingerJK commented on changelog/README.md:
```diff
@@ -0,0 +1,59 @@
+
+To avoid merge conflicts in the CHANGELOG.md file we accepted the multi-file running changelog strategy.
+
+To follow this strategy you should create a `changelog/unreleased/${PR_NUMBER}.json` file for every PR like:
```
for that approach, we have to add a changelog in another commit (first create PR, get the PR number, and put the changelog and PR number into the branch).

Can it be modified so that we don't add the PR number manually? Can it be added under the hood when PR has been created?

dzinad replied:
> for that approach, we have to add a changelog in another commit

I like to use the `-f` option but yes, I agree that it can get annoying.

SevaZhukov replied:
I understand that it's not comfortable to write a changelog only after creating PR. I can propose creating an action that will rename the changelog file to ${PR_NUMBER}.md:
1. You create a PR (with changelog file feature_1.md)
2. The action executes and rename the changelog file to ${PR_NUMBER}.md
Action will be executed only on open/reopen/redy for review moments
WDYT

RingerJK replied:
lgtm 👍 

how does it works for release branches where we mostly cherry-pick changes (sometimes already released)? ref https://github.com/mapbox/mapbox-navigation-android/pull/6740#issuecomment-1352894603

SevaZhukov replied:
oh it's a good moment, I think we can create a PR to the release branch first. Then, the action will rename the changelog file. Then, we can cherry-pick commits (changes + renaming commit) to the main branch

kmadsen replied:
> oh it's a good moment, I think we can create a PR to the release branch first. Then, the action will rename the changelog file. Then, we can cherry-pick commits (changes + renaming commit) to the main branch

Typically we merge into main branch, and then cherry-pick into the release branch(es). Maybe minor detail but I don't think we want it to require release commits first.

RingerJK commented:
@SevaZhukov would be any corner cases for release branches? Could you mention if any in `readme`

SevaZhukov commented:
> Should we let to insert sections like [this](https://github.com/mapbox/mapbox-navigation-android/blob/main/CHANGELOG.md#known-issues-warning) or it should be done manually during release?

Good point, I have added this section

dzinad commented on changelog/README.md:
````diff
@@ -0,0 +1,58 @@
+```
+
+You can choose any name for your changelog files because the GitHub action will rename files in
+`changelog/unreleased/features` and `changelog/unreleased/bugfixes` directories to `${PR_NUMBER}.md` when you open a PR.
````
What about the issues directory and just `changelog/unreleased`?

SevaZhukov replied:
we need PR's links only for bugfixes and features

dzinad commented on changelog/README.md:
````diff
@@ -0,0 +1,58 @@
+```
+
+* write the changelog to the `CHANGELOG.md` file
+* delete all files from `changelog/unreleased`
````
Just make sure patch releases won't erase all main changelog entries...

SevaZhukov replied:
Releases since rc.1 will erase changelog/unreleased dir only in release branches

dzinad commented on changelog/unreleased/bugfixes/6540.md:
```diff
@@ -0,0 +1 @@
+Fixed approaches list update in `RouteOptionsUpdater`(uses for reroute). It was putting to the origin approach corresponding approach from legacy approach list.
```
Shouldn't there be a `-` in the beginning?

SevaZhukov replied:
I think the script will add `-` during assembling

dzinad replied:
It's not a very good idea. If the entry is multi-line, how will the script understand where to add the `-`?

SevaZhukov replied:
if you have only one feture/bugfix you can don't use `-` before, if you have more you should

dzinad commented on changelog/unreleased/bugfixes/6695.md:
```diff
@@ -0,0 +1 @@
+Updated the `MapboxRestAreaApi` logic to load a SAPA map only if the upcoming rest stop is at the current step of the route leg.
```
Shouldn't there be a `-` in the beginning?

dzinad commented on changelog/unreleased/issues/example-known-issues.md:
```diff
@@ -0,0 +1 @@
+It is an example of known issues
```
Shouldn't there be a `-` in the beginning?

dzinad commented on scripts/validate_changelog_utils.py:
```diff
@@ -7,15 +7,19 @@
 diff_file_start_regex = "^([\s]*)\@\@(.*)\@\@"
 pr_link_regex = "\[#\d+]\(https:\/\/github\.com\/mapbox\/mapbox-navigation-android\/pull\/\d+\)"
 
+
```
That's not what I meant when saying we should adapt the checks...

SevaZhukov replied:
I didn't do anything with it))

dzinad replied:
But there are changes in the diff...

SevaZhukov replied:
only autoformating

dzinad commented on changelog/README.md:
````diff
@@ -0,0 +1,58 @@
+- Description of changes in md format also
+```
+
+You can choose any name for your changelog files because the GitHub action will rename files in
````
I think it's worth mentioning that you can't create 2 files in 1 directory in 1 PR. I don't think we someone would want that (I don't see any cases for that) but still we should note that.

SevaZhukov replied:
but I mentioned it [here](https://github.com/mapbox/mapbox-navigation-android/pull/6740/files/0df6090a3f31bb3559f877434447707d953af319#diff-861d0d6527afefde8a5f11cd0baad47e651931f7424b83462f0f66875afde016R14)

SevaZhukov commented:
hey team, I added two GitHub actions
1. Rename changelog files. It executes every update and looking for and renaming new changelog files in `features` and `bugfixes` dirs from `any-feature-name.md` to `${PR_NUMBER}.md`. After it pushes to the current branch.
2. Assemble changelog. It executes only on the main and release branches. This action compile changelog from `features`, `bugfixes` and `issues` dirs like:
```
#### Features
- Test changes to check renaming
- Introduced `MapboxSpeedInfoApi` and `MapboxSpeedInfoView`. The combination of API and View can be used to render posted and current speed limit at user's current location.
- :warning: Deprecated `MapboxSpeedLimitApi` and `MapboxSpeedLimitView`.

#### Bug fixes and improvements
- Updated the `MapboxRestAreaApi` logic to load a SAPA map only if the upcoming rest stop is at the current step of the route leg.
- Fixed approaches list update in `RouteOptionsUpdater`(uses for reroute). It was putting to the origin approach corresponding approach from legacy approach list.

#### Known issues :warning:
- It is an example of known issues
```
 After it commits and pushes to the current branch. Commit name contains `[skip actions]` label to skip this commit for the actions.

@tomaszrybakiewicz @kmadsen @VysotskiVadim @RingerJK @dzinad

dzinad commented on changelog/README.md:
```diff
@@ -0,0 +1,58 @@
+
+You can use anything that allow .md format in changelog files.
+
+If you have implemented some features or bugfixes you should describe all of them:
```
some -> several/multiple

dzinad commented on changelog/README.md:
```diff
@@ -0,0 +1,58 @@
+You can choose any name for your changelog files because the GitHub action will rename files in
+`changelog/unreleased/features` and `changelog/unreleased/bugfixes` directories to `${PR_NUMBER}.md` when you open a PR.
+
+Every release the release train app will:
```
It's not only on every release, right? It will happen after every merge? Can you actualize README? Because that's all I'm reviewing actually: this file and the examples. :)

SevaZhukov replied:
updated

dzinad commented on changelog/unreleased/CHANGELOG.md:
```diff
@@ -0,0 +1,11 @@
+#### Features
+- Test changes to check renaming
```
What about the PR links? They should be here because we've lost all the info about where this entry came from at this point, right?

SevaZhukov replied:
I added links to assembled changelog, thanks

dzinad commented on changelog/README.md:
````diff
@@ -0,0 +1,45 @@
+Some other changes
+```
+
+* write the changelog to the `changelog/unreleased/CHANGELOG.md` file
````
What about the release process? Shouldn't we move the changelog from unreleased directory to the project root and delete the unreleased one?

SevaZhukov replied:
Added more information

dzinad commented on changelog/README.md:
```diff
@@ -0,0 +1,45 @@
+You can choose any name for your changelog files because the GitHub action will rename files in
+`changelog/unreleased/features` and `changelog/unreleased/bugfixes` directories to `${PR_NUMBER}.md` when you open a PR.
+
+Every push to the main or release branch Assemble changelog GitHub action will be executed:
```
The temporary files will be deleted, right?

SevaZhukov replied:
no, let's keep they before the release

dzinad commented:
The docs and the examples look good. And I believe you know what you're doing with the scripts. :)

kmadsen commented on changelog/README.md:
````diff
@@ -0,0 +1,79 @@
+- Mapbox Android Core `v5.0.2` ([release notes](https://github.com/mapbox/mapbox-events-android/releases/tag/core-5.0.2))
+```
+* add compiled changelog to `CHANGELOG.md` file
+* delete all files in `changelog/unreleased` dir
````
@SevaZhukov will this system work with the android auto changelog https://github.com/mapbox/mapbox-navigation-android/blob/main/libnavui-androidauto/CHANGELOG.md

All good to consider it out of scope but it would be nice if it supported multiple release channels

VysotskiVadim commented on changelog/README.md:
```diff
@@ -0,0 +1,81 @@
+You can choose any name for your changelog files because the GitHub action will rename files in
+`changelog/unreleased/features` and `changelog/unreleased/bugfixes` directories to `${PR_NUMBER}.md` when you open a PR.
+
+Every push to the main or release branch Assemble changelog GitHub action will be executed:
```
Will those commit trigger metrics run? For now it seems so, because each PR merge produces 2 commits in the main branch.

SevaZhukov replied:
[skip ci] tag allows skipping CircleCI jobs and GitHub Actions

VysotskiVadim commented on scripts/changelog/assemble_changelog.py:
```diff
@@ -0,0 +1,60 @@
+import git
+
+
+def get_changes(path):
```
What is the testing strategy for the changelog logic? Do I understand correctly that after each change I need to manually run script locally simulating different scenarios that my change may affect? Do you think it makes sense to add auto tests, so that developers could enhance this logic without being afraid break a different scenario?

I think this question isn't a blocker for this particular PR, but it's worth to discuss 🙂

SevaZhukov replied:
yeah, good point, let's add tests for this logic

RingerJK commented on changelog/README.md:
```diff
@@ -0,0 +1,81 @@
+
+To follow this strategy you should create a `.md` file for every PR. Choose a directory:
+
+- `changelog/unreleased/features` for **Features** changes
```
IMO would be nice to have a script to add changelog like `python add_changelog.py -f "Changelog string"`, where `-f` is `feature` (also should be other flags), `"Changelog string"` is the entry of changelog

SevaZhukov replied:
done
